custom_components/elasticsearch/es_index_manager.py

```diff
@@ -69,9 +126,20 @@ async def _create_index_template(self):
 
         LOGGER.debug('checking if template exists')
 
-        template = await client.indices.get_template(name=INDEX_TEMPLATE_NAME, ignore=[404])
+
+        # check for 410 return code to detect serverless environment
+        try:
+            template = await client.indices.get_template(name=LEGACY_TEMPLATE_NAME, ignore=[404])
+
+        except ElasticsearchException as err:
+            if err.status_code == 410:
+                LOGGER.error(
+                    "Serverless environment detected, legacy index usage not allowed in ES Serverless. Switch to datastreams."
+                )
+                raise err
```

Comment:
This comment is optional, we can explore feasibility in a followup if you'd like.

We may be able to detect this condition earlier in the flow by updating the `ElasticsearchVersion` class to expose an `is_serverless` property (or similar). We could then prevent the user from selecting the "legacy" index mode during setup if we detect they're running in a serverless environment.

https://github.com/legrego/homeassistant-elasticsearch/blob/8e6ac2d91868b4e2e5dbf1687035c9ff503b496b/custom_components/elasticsearch/es_version.py#L4

Reply:
I opened https://github.com/legrego/homeassistant-elasticsearch/pull/211 to add this support to `ElasticsearchVersion`

Reply:
I will move it "earlier in the flow" by moving it up like ten lines but moving it to the config will take me a bit longer so I'll leave this comment open for now